### PR TITLE
Fix NPEs in ModificationsSyncAdapter

### PR DESCRIPTION
**Pull request by misaochan**
For #1087 . Adds null check at crash location based on @psh 's recommendation. Also adds a few other null checks to areas where Android Studio flagged potential NPEs. Please let me know if I handled the null checks appropriately.

A TODO that help would be appreciated with - best way to prevent NPE for a cursor query?

Also, we probably need to reexamine why the cursor is empty to begin with, in addition to these fixes.

### CODE REVIEW

**Comment by whym** [comment]
Looks good to me. Upload went through for me, but I had't experienced the crash, so this is kind of expected.

> Also, we probably need to reexamine why the cursor is empty to begin with

Any idea on helping people report this issue without making them experiencing crashes?

We could perhaps add a log message when a null is detected in one of those variables. (Well, it may not be much of help, but at least people who can use logcat can help us.)

**Comment by misaochan** [comment]
@whym Thanks for the feedback! Let's carry on the discussion in the issue for continuity, I'll copy your post over there. :)

**Comment by neslihanturan** [comment]
@misaochan Actually I don't know how to review this PR since I never reproduced the bug.

**Comment by misaochan** [comment]
@neslihanturan Yes, unfortunately none of us have managed to reproduce the bug. :( However, if you test it and it doesn't break anything for you, I guess we can merge it while keeping #1087 open to carry on the discussion about the empty cursor.

**Comment by misaochan** [comment]
BTW, when this is tested, please make sure that the test upload has categories added (as AFAIK that is the  purpose of ModificationsSyncAdapter). Without any category selections, I don't think this class will be called.

**Comment by neslihanturan** [comment]
Hi team, I am able to reproduce the bug with devices 5.0.2 and 4.42 API. It happens with category and without category uploads. When I follow debug break points, weirdly ModificationSyncAdapter is called even if I don't add any categories. 

- I am able to upload with web site and see my new uploads from app's interface.
- When I upload an image with category it is normal ModificationSyncAdapter is being called. But according to my debug, it is called on uploads without categories and even with no uploads. And here is @misaochan 's great theory comes, whenever a category of previous uploads didn't added, ModificationSysncAdapter being called to add this missing category. It was a known issue #310 . But currently it may trigger NPE. 
- I was able to upload while testing our master before release, but at some point I couldn't be able to upload anymore (and I don't think this is API level dependent), guess what one of my category information is not added yet.

File: app/src/main/java/fr/free/nrw/commons/modifications/ModificationsSyncAdapter.java [hunk 1]
```diff
@@ -95,19 +95,24 @@ public void onPerformSync(Account account, Bundle bundle, String s, ContentProvi
 
                 Cursor contributionCursor;
                 try {
+                    //TODO: How do we prevent this NPE?
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/modifications/ModificationsSyncAdapter.java]: return early if `contributionsClient` is null.

- also add a null check for getContext()

**Comment by misaochan** [comment]
According to @neslihanturan , when testing the effect of this PR on the bug, even though this PR prevents the app from crashing, all of her subsequent uploads fail regardless.

In that case I guess I'm not sure if these fixes will be beneficial. At least when the app crashes, we get a notification in the Play Store dev console that tells us something is going wrong. On the other hand, if the app is silently just absorbing the empty cursor but still preventing the user from uploading, that may be worse, as it will be unusable for the user but we get no information.

**Comment by neslihanturan** [comment]
I also think this PR is not what we needed. We need to remove root cause of our headache, not painkilling it:)

**Comment by misaochan** [comment]
I agree @neslihanturan . Will close this for now. :)

**Comment by misaochan** [comment]
Reopening based on discussion at #1087 , so it is easier for others to review this. I can put in the changes suggested by @maskaravivek if this (Option 3) is the temporary solution that we decide on.

**Comment by neslihanturan** [comment]
Since our lack of understanding of our bug, I don't feel confident to silence the bug reports with this PR. Can we still get information whenever this cursor is null without violating users privacy when we remove the crash with this PR?

**Comment by misaochan** [comment]
Reposting from Hangouts: 

My thoughts are that since future uploads do not actually fail, there is no harm in silencing the crash for the time being. Given a choice between ONE upload failing and a crash, I think one upload failing is less disruptive to the user

However, we should keep the issue running to see if the empty cursor reason can be found. When we find it and try to fix it, we can remove the null check from ModificationsSyncAdapter to see if our fix works. But I don't see any reason to leave the crash up until that time comes.

File: app/src/main/java/fr/free/nrw/commons/modifications/ModificationsSyncAdapter.java (hunk 1 above)
Comment (misaochan) [reply app/src/main/java/fr/free/nrw/commons/modifications/ModificationsSyncAdapter.java]: Thanks, I have added the check for contributionsClient. :) However getContext() does not raise any possible NPE flags in Android Studio, is a null check really necessary whenever we call that?

**Comment by misaochan** [comment]
Added error logging, the final null check, and switched base (to be in line with the system mentioned at #1115 ).

**Comment by misaochan** [comment]
As this is an urgent fix (current crash rate 13.7%), I am just going to go ahead and merge into the release branch. The issue will remain open for discussion and future fixes. :)

**Comment by nicolas-raoul** [comment]
This is where the last sentence of https://github.com/commons-app/apps-android-commons/wiki/Project-maintenance#pull-requests-and-merging-workflow comes in handy :-)

In another project I used ACRA (open source crash reporting library+analysis webapp) to report not only crashes but also "should never happen" situations like this. This could be useful here.

I just tested and my upload went through normally (I had no crash before either, though)

**Pull request merged**